File: 백준/Gold/1987. 알파벳/알파벳.java

import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class Main {
	static String[][] grid;
	static boolean[] visited = new boolean[26];
	static int R;
	static int N;
	static int result;
	static int dx[] = {-1, 1, 0, 0};
	static int dy[] = {0, 0, -1, 1};
	
	public static void main(String[] args) {
		Scanner scan = new Scanner(System.in);
		
		R = scan.nextInt();
		N = scan.nextInt();
		
		scan.nextLine();
		
		grid = new String[R][N];
		
		for(int i = 0; i < R; i++) {
			String[] temp = scan.nextLine().split("");
			for(int j = 0; j < N; j++) {
				grid[i][j] = temp[j];
			}
		}
		
		result = 0;
		alpha(0, 0, 0);
		
		System.out.println(result);
	}
	
	public static void alpha(int x, int y, int count) {
		if(visited[grid[x][y].charAt(0) - 'A']) {
			result = Math.max(result,  count);
			return;
		}
		else {
			visited[grid[x][y].charAt(0) - 'A'] = true;
			for(int i = 0; i < 4; i++) {
				int nx = x + dx[i];
				int ny = y + dy[i];
				
				if(nx >= 0 && nx < R && ny >= 0 && ny < N) {
					alpha(nx, ny, count + 1);
				}
			}
			
			visited[grid[x][y].charAt(0) - 'A'] = false;
			
		}
	}
	
	
	
}

File: 백준/Gold/1987. 알파벳/알파벳.py
R, C = map(int, input().split())

graph = []
for _ in range(R):
    graph.append(list(input()))

visited = [False] * 26

result = 0
dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]

def dfs(x, y, count):
    if visited[ord(graph[x][y]) - ord('A')]:
        global result
        result = max(result, count)
    else:
        visited[ord(graph[x][y]) - ord('A')] = True
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]

            if 0 <= nx < R and 0 <= ny < C:
                dfs(nx, ny, count + 1)

        visited[ord(graph[x][y]) - ord('A')] = False

dfs(0, 0, 0)

print(result)